Auto collect fresh data to `collect-data` branch

twagner000 opened the pull request:
- Use workflow to collect all non-webkit states every Monday and push to `collect-data` branch of this repo
- Increase wait delay for Virginia to avoid Access Denied errors
- Update `test_michigan_fips.py` to skip test if michigan.json doesn't exist
- Ran into issues with qt for `cache_webkit` (only used by Nevada). Tried using `platform offscreen` through env variable QT_QPA_PLATFORM and/or arg to `QApplication()`. Also tried using xvfb-run.  Unable to resolve because site started asking for captcha (modified `cache_webkit` to print out html to see if running into a captcha). Created a separate workflow for webkit states for future improvement.

twagner000 committed:
collect alaska, wyoming and commit to `collect-data` branch

twagner000 committed:
use 2 repos in separate dirs

- cd for each action
- add directory for push

twagner000 committed:
increase virginia delay to avoid Access Denied errors

twagner000 committed:
debug nevada (cache_webkit) qt issues

Ran into issues with qt for `cache_webkit`. Tried using `platform offscreen` through env variable QT_QPA_PLATFORM and/or arg to `QApplication()`. Also tried using xvfb-run. Unable to resolve because site started asking for captcha.

twagner000 committed:
only test_michigan_fips if michigan.json exists

twagner000 committed:
split into webkit and no-webkit states

tianhuil commented on .github/workflows/collect-webkit.yml:
```diff
@@ -0,0 +1,64 @@
+        python-version: 3.7
+    - name: Create venv and install dependencies
+      run: |
+        pwd
```
I don't think yo need `pwd` here and below.

twagner000 replied:
Agreed, they're remnants of development. Cleaned up to remove unnecessary `pwd` and `ls` commands.

tianhuil commented on .github/workflows/collect-webkit.yml:
```diff
@@ -0,0 +1,64 @@
+      run: |
+        pwd
+        cd code_repo
+        ls -ls public
```
`ls -ls public` isn't necessary either

tianhuil commented on .github/workflows/collect-webkit.yml:
```diff
@@ -0,0 +1,64 @@
+      uses: actions/checkout@v2
+      with:
+        path: 'data_repo'
+        ref: 'collect-data'
```
Can we make this branch `public-data`?

tianhuil commented on .github/workflows/collect-webkit.yml:
```diff
@@ -0,0 +1,64 @@
+        git config --local user.name "GitHub Action"
+        git add ./public
+        git commit -m "Collect new data files"
+        git push origin collect-data
```
Can we add a tag with the date?  We want a stable reference as well as a reference to 'latest' (this branch).

tianhuil commented on .github/workflows/collect.yml:
```diff
@@ -0,0 +1,61 @@
+# This workflow will install Python dependencies, collect data for states that DON'T use webkit, and run tests on the generated files
```
This is fine.  Let's merge these eventually.

twagner000 replied:
Agreed, as soon as we can get the webkit workflow working consistently they should be merged.

tianhuil commented on src/common.py:
```diff
@@ -90,7 +90,9 @@ def _load_finished(self, result):  # pylint: disable=unused-argument
     time.sleep(wait)
-  return Render(url).html
+  html = Render(url).html
+  print(html)
```
In general, we should remove print statements

twagner000 replied:
I moved the print statement, with an explanatory comment about the occasional captcha, to Nevada's `fetch_data()`

tianhuil commented on src/tests/test_michigan_fips.py:
```diff
@@ -18,9 +18,11 @@ def get_arcgis_fipscode():
+    with open(file_) as json_file:
+      rows = json.load(json_file)
+      return sorted([row['fipscode'] for row in rows])
+  return None
```
Consider throwing an error if the file is not found.

twagner000 replied:
An error can prevent other, successful updates from getting pushed through. E.g., the webkit workflow which only runs Nevada (so michigan.json doesn't exist).

tianhuil replied:
Good call.

tianhuil reviewed (changes requested):
A few changes.  Would recommend removing `public/` via .gitignore from `master` but keeping it in the `public-data` branch.

twagner000 committed:
add .gitignore to `public/` for master

- NOTE: `public/` should NOT have .gitignore for `public-data` branch since this is where we plan to actually keep the data.
- No longer need to clear `public/` in workflows

twagner000 committed:
remove unnecessary commands; change data branch to `public-data`

twagner000 committed:
list .json files tested

twagner000 committed:
remove print statement from `cache_webkit`

move into nevada `fetch_data()` with a comment where the captcha sometimes occurs

tianhuil reviewed (changes requested):
Basically good.

1. Can you write up documentation in `README.md` for how to get the latest version URL?  Make sure we continue to document the old way as well.
2. Can you resolve the conflict we introduced through NH?
3. Want to make 100% sure that the old versions (e.g. everything through `1.7.1`) don't get deleted.
4. Remove code from public data branch (but keep data)
5. Add README.md to the public data branch
6. Push a working copy of all data with a tag onto the public-data branch
7. Give instructions for how to manually update the public-data branch

twagner000 committed:
tag public data with `data/YYYY-MM-DD`

twagner000 committed:
test selected states

twagner000 committed:
update README with data changes

twagner000 committed:
move test to separate workflow since no files in `public/`

twagner000 committed:
add `limited` states, print list of json files tested

twagner000 committed:
CI - cache requests and dependencies

- run all states in `limited` using cached requests (if available)
- request cache uses `restore-keys` to load the most recent existing cache (***NOT*** used for public-data)
- request cache uses a datetime key to always save a fresh copy, in case additional files are downloaded
- use of `always-upload-cache` means cache is still saved in case of errors
- resulting json files are uploaded as an artifact
- public-data scheduled to run on Friday using `cron`

twagner000 committed:
virginia - attempt retry if error

also, update virginia wait time

twagner000 committed:
remove initial workflow attempts

twagner000 committed:
comment why nevada is in SKIP_STATES

twagner000 committed:
Update badges

twagner000 committed:
remove `restore-keys` from public-data workflow to force collection of fresh data

twagner000 commented:
New approach is ready for review. Used cached requests for CI testing. Weekly scheduled public data workflow will start a new requests cache. Nevada is still excluded from auto update for now due to captcha issues.

tianhuil commented on src/states/virginia/main.py:
```diff
@@ -12,30 +15,40 @@ def get_locality_ids():
+  except IndexError:
+    cached_files = glob.glob(os.path.join(work_dir, '*.*'))
+    latest_file = max(cached_files, key=os.path.getctime)
+    print(f"error in Virginia file; deleting cached file {latest_file}; retry after wait")
```
Instead of print for logging, it's preferred to use `logging.error`, `logging.warn`.

tianhuil commented on README.md:
````diff
@@ -45,6 +50,7 @@ Each state's crawler is put under its own folder (e.g. `states/new_york/main.py`
   ```make test```
 
 - Also, rerun the Jupyter notebook `analysis/Analysis.ipynb` from scratch to update the analytics.  You can see how many fields you were able to parse.  To start the jupyter notebook, run `make jupyter`.  Run the notebook.  Make sure that you have all the values you need.  **Do not commit the notebook changes**.  Jsut throw them away.  They just block rebase merging.
+- To release a new data version between scheduled run dates, ***first make sure the data file passes all tests*** using `inv test`. Then, commit the passing data file to the [public-data branch](https://github.com/mail-my-ballot/elections-officials/tree/public-data) and tag that commit with a date version of format `data/2020-06-22`.
````
Is there an `inv` workflow that could do these steps? so that you could say run `inv collect`

twagner000 replied:
I'm a bit wary of adding an inv task that does that many git actions. It would involve:

1. running a state
2. running test
3. copying the .json file somewhere
4. switching branches to public-data
5. fetch/pull
6. copy .json file back to public
7. add .json file
8. commit (if different)
9. tag (if different)

I think there are too many things that could go wrong and inadvertently really mess up someone's local repo.

Perhaps we could come up with an additional trigger to just run the public_data workflow on demand in those situations?

twagner000 replied:
Perhaps an inv task that submits a POST request as an [external repository dispatch](https://help.github.com/en/actions/reference/events-that-trigger-workflows#external-events-repository_dispatch) to the public_html workflow? Or to a new workflow if we want it to be state-specific and pass the state as part of the [client payload](https://developer.github.com/v3/repos/#create-a-repository-dispatch-event)?

Seems like a lot of effort for not much gain, but it is possible.

tianhuil replied:
OK -- let's merge for now and discuss this somewhere else.

tianhuil commented on src/states/nevada/main.py:
```diff
@@ -37,6 +37,10 @@ def parse_county(text):
   html = cache_webkit(BASE_URL)
+
+  # watch for occasional captcha requirement (no current way around it)
+  print(html)
```
`print` should be `logging.warn` or something like this.

tianhuil commented on .github/workflows/python_ci.yml:
```diff
@@ -0,0 +1,48 @@
+      with:
+        path: ~/.cache/pip
+        key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements.txt') }}
+        restore-keys: |
```
Let's remove restore-keys here

tianhuil commented on .github/workflows/public_data.yml:
```diff
@@ -0,0 +1,72 @@
+        git config --local user.name "GitHub Action"
+        git commit -m "Collect new data files $COMMIT_DATE"
+        git push origin public-data
+        git tag data/$COMMIT_DATE
```
Push an updated tag to the same commit even if no new commit entered.

tianhuil commented on .github/workflows/public_data.yml:
```diff
@@ -0,0 +1,72 @@
+      run: make test
+    - name: upload artifact public/*.json
+      uses: actions/upload-artifact@v2
+      with:
```
Add in note about how we can examine artifacts for debugging

tianhuil commented on src/states/virginia/main.py:
```diff
@@ -12,30 +15,40 @@ def get_locality_ids():
+      'physicalAddress': results.get('Physical Address'),
+    }
+    return {k: v for k, v in final.items() if v}
+  except IndexError:
```
Instead of using a general `except`, check `if soup.select('.resultsWrapper')` instead

tianhuil commented on .github/workflows/public_data.yml:
```diff
@@ -0,0 +1,72 @@
+
+on:
+  schedule:
+    - cron: '0 0 * * 5'
```
Let's make this every day to start out.

twagner000 committed:
use logging instead of print

twagner000 committed:
more specific handling `.resultsWrapper` not found

twagner000 committed:
don't use backup cache for dependencies (no `restore-keys`)

twagner000 committed:
comments, tag data if no changes, start scheduled daily

- comment workflow files
- tag data even if no changes
- start scheduled every day (follow up to change to once per week once in master)

twagner000 committed:
tmp changes to test public_data updates

twagner000 committed:
lint fixes

twagner000 committed:
Revert "tmp changes to test public_data updates"

This reverts commit f72d18c1a8355938fd7b4e894d54a20dab01dd0d.

twagner000 committed:
add blank line to requirements.txt to reset dependencies cache

twagner000 committed:
comments

twagner000 commented:
All updates made except for `inv` action to update public-data between scheduled runs (see comments above).
